drivers/mbox/mbox_andes_plic_sw.c

```diff
@@ -0,0 +1,245 @@
+	}
+
+	if (!is_channel_valid(dev, ch)) {
+		return -EINVAL;
+	}
+
+	/* Send IPI by triggering the pending register of PLIC SW. */
+	plic_sw_irq_set_pending(dev, ch + 1);
+
+	return 0;
+}
+
+static int mbox_andes_register_callback(const struct device *dev, uint32_t ch,
+					mbox_callback_t cb, void *user_data)
+{
+	struct mbox_andes_data *data = dev->data;
+	const struct mbox_andes_conf *conf = dev->config;
+	int ret = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&mbox_syn);
+
+	if (ch > conf->channel_max) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+#ifdef CONFIG_SCHED_IPI_SUPPORTED
+	if (ch & data->ipi_channel & data->reg_cb_channel) {
+		ret = -EALREADY;
+		goto out;
+	}
+
+	data->reg_cb_channel |= BIT(ch);
+#endif
+	data->cb[ch] = cb;
+	data->user_data[ch] = user_data;
+
+out:
+	k_spin_unlock(&mbox_syn, key);
+
+	return 0;
+}
+
+static int mbox_andes_mtu_get(const struct device *dev)
+{
+	/* We only support signalling */
+	return 0;
+}
+
+static uint32_t mbox_andes_max_channels_get(const struct device *dev)
+{
+	const struct mbox_andes_conf *conf = dev->config;
+
+	return conf->channel_max;
+}
+
+static int mbox_andes_set_enabled(const struct device *dev, uint32_t ch,
+				  bool enable)
+{
+	uint32_t en, is_enabled_ch, hartid, cpu_id, irq;
+	struct mbox_andes_data *data = dev->data;
+	int ret = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&mbox_syn);
+
+	if (!is_channel_valid(dev, ch)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	irq = ch + 1;
+	hartid = arch_proc_id();
+	cpu_id = _current_cpu->id;
+
+	is_enabled_ch = data->enabled_channel[cpu_id] & BIT(ch);
+
+	if ((!enable && !is_enabled_ch) || (enable && is_enabled_ch)) {
+		ret = -EALREADY;
+		goto out;
+	}
+
+	if (enable && !(data->cb[ch])) {
+		LOG_WRN("Enabling channel without a registered callback\n");
+	}
+
+	en = sys_read32(REG_ENABLE(dev, hartid, irq));
+
+	if (enable) {
+		data->enabled_channel[cpu_id] |= BIT(ch);
+		sys_write32(1, REG_PRIORITY(dev, irq));
+		en |= BIT(irq);
+	} else {
+		data->enabled_channel[cpu_id] &= ~BIT(ch);
+		en &= ~BIT(irq);
+	}
+
+	sys_write32(en, REG_ENABLE(dev, hartid, irq));
+out:
+	k_spin_unlock(&mbox_syn, key);
+
+	return ret;
+}
+
+static void andes_plic_sw_irq_handler(const struct device *dev)
+{
+	struct mbox_andes_data *data = dev->data;
+	uint32_t irq, ch, hartid;
+
+	hartid = arch_proc_id();
+
+	/* PLIC claim: Get the SW IRQ number generating the interrupt. */
+	irq = sys_read32(REG_CLAIM(dev, hartid));
+	ch = irq - 1;
+
+	if (irq) {
+		sys_write32(irq, REG_CLAIM(dev, hartid));
+
+		if (data->cb[ch]) {
+			/* Only one MAILBOX, id is unused and set to 0 */
+			data->cb[ch](dev, ch, data->user_data[ch], NULL);
+		}
+	}
+}
+
+static int mbox_andes_init(const struct device *dev)
+{
+#ifdef CONFIG_SCHED_IPI_SUPPORTED
+	uint32_t ipi_num;
+	struct mbox_andes_data *data = dev->data;
+
+	static const struct mbox_channel ipi_channel[] = {
+		DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(mbox), mbox_names,
+		IPI_MBOX_CHANNEL, (,))
+	};
```

Comment:
what are you trying to do with this? To me it looks like this is always an empty array because `IPI_MBOX_CHANNEL()` is never actually binding to anything.

Reply:
IPI_MBOX_CHANNEL is for the future using of smp ipi. after discussing, these code is not appropriate upstream now, we have removed it.
Thanks for your comment.